pkg/reconciler/managed/reconciler.go

```diff
@@ -1159,45 +1154,48 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (resu
 		log.Debug("External resource differs from desired state", "diff", observation.Diff)
 	}
 
-	// skip the update if the management policy is set to ignore updates
-	if !policy.ShouldUpdate() {
-		reconcileAfter := r.pollIntervalHook(managed, r.pollInterval)
-		log.Debug("Skipping update due to managementPolicies. Reconciliation succeeded", "requeue-after", time.Now().Add(reconcileAfter))
-		managed.SetConditions(xpv1.ReconcileSuccess())
-		return reconcile.Result{RequeueAfter: reconcileAfter}, errors.Wrap(r.client.Status().Update(ctx, managed), errUpdateManagedStatus)
-	}
+	if !observation.ResourceUpToDate && policy.ShouldUpdate() {
```

Comment:
This if condition is what allows us to treat the `Update` case the same as the `Create` case.

This is as close to the implementation for `Create` as possible.

Reply:
Note that this reconciler sets the `Creating` condition, but it never unsets it. It never sets the `Available` condition.

The reconciler can know that it triggered the creation of of an external resource, but it has no way to know whether creation is done. The `ExternalClient` is responsible for setting the resource to `Available` when it notices that creation has completed. For some resources the `ExternalClient` knows that creation has completed due to some API level state transition. Other resources are ready as soon as they're created.


Will `ExternalClient` implementations need to factor in the existence of a new `Updating` condition by handling the case where a resource has transitioned from `Updating` back to `Available`?

Reply:
> You seem to focus a lot on what the cloud provider reports for the resource, so if they say it is ready, we set it to ready, but is this really what we want?

Today we have the following status conditions:

* `Ready` - The external resource is ready to use, (where applicable) according to its API.
* `Synced` - The managed resource's desired state has been successfully applied to the external resource.

I don't think we can change their meaning at this point.

I do agree there's a gap though. `Synced` tells you the latest desired state was _successfully submitted_ to the external API, but there's no condition that is a positive confirmation that the external resource's observed state has converged on the latest desired state.

This would be similar to to the `Current` status in kstatus, per https://github.com/kubernetes-sigs/cli-utils/tree/master/pkg/kstatus. Though note that's not a status condition but rather a status field.

Some related issues:

* https://github.com/crossplane/crossplane/issues/2672
* https://github.com/crossplane/crossplane/issues/4655
* https://github.com/crossplane/crossplane/issues/4695

Reply:
I wasn't actually able to find these definitions in the docs, only this from the troubleshooting part:
```
Most Crossplane resources set the Ready condition. Ready represents the availability of the resource - whether it is creating, deleting, available, unavailable, binding, etc.
```

You acknowledge that there is a gap here, and I feel like the issues you linked confirm the gap and that is a big enough problem that workarounds have been attempted to fix it.

I understand that definitions and such can be scary to change, as they often introduce some level of a breaking change, but I don't really see the negative side of changing this. All providers already implement a way of checking if the resource is actually compliant with the desired state, I only suggest that we actually use it for `Update` too.

What is the argument for not changing this?
The only negative I see is that the resource will be shown as `Updating` for a couple of seconds before it shows `Ready`.

Another issue with the current implementation: 
If I were to create my own in-house provider, and I wanted the `Ready` field to better indicate if the resource is updating, I could manually set the condition in my `Update` function. The problem is that there is no reconcile requeue after successfully requesting the update, so if I have 10 minute poll interval, it means my resource is `Updating` for 10 minutes, even though it only took a second.

Reply:
> What is the argument for not changing this?

I'm not convinced the way you're addressing the problem in this PR is the best way to address the problem. If we're going to change the semantics of `Ready`, we should take a step back and:

* Consider the impact.
* Consider whether changing the semantics of `Ready` is the right thing to do (vs alternatives like adopting `kstatus`.

I think a change like this warrants at least a one-pager.

Reply:
Sounds good! Would you want me to write it, or is this something you would do internally?
I am not very familiar with the `one-pager` concept, but it seems like most of the existing ones are written by maintainers.

Reply:
I'm definitely happy for you to write it. 😄 I think it's a great thing to have more community members leading designs. We can pair you up with either myself or a maintainer closer to your timezone (e.g. @phisco or @turkenh) to help bounce ideas from.

Reply:
Sure, I can give it a try 😄 
I'll work on a first draft and open a PR in the Crossplane repo. I will do some research into Kstatus too.